File: android/app/src/main/java/com/tesla254/Textly/SmsReceiver.java
package com.tesla254.Textly;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.provider.Telephony;
import android.telephony.SmsMessage;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

public class SmsReceiver extends BroadcastReceiver {
    private static final String TAG = "SmsReceiver";
    private ReactApplicationContext reactContext;

    public SmsReceiver(ReactApplicationContext reactContext) {
        this.reactContext = reactContext;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent.getAction().equals(Telephony.Sms.Intents.SMS_RECEIVED_ACTION)) {
            Bundle bundle = intent.getExtras();
            if (bundle != null) {
                Object[] pdus = (Object[]) bundle.get("pdus");
                if (pdus != null) {
                    for (Object pdu : pdus) {
                        SmsMessage smsMessage = SmsMessage.createFromPdu((byte[]) pdu);
                        
                        WritableMap params = Arguments.createMap();
                        params.putString("address", smsMessage.getDisplayOriginatingAddress());
                        params.putString("body", smsMessage.getDisplayMessageBody());
                        params.putDouble("date", smsMessage.getTimestampMillis());
                        params.putInt("simSlot", getSimSlot(smsMessage));
                        
                        // Send event to React Native
                        sendEvent("onSmsReceived", params);
                        
                        Log.d(TAG, "SMS received from: " + smsMessage.getDisplayOriginatingAddress());
                    }
                }
            }
        }
    }

    private int getSimSlot(SmsMessage smsMessage) {
        try {
            // Try to get SIM slot from message
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP_MR1) {
                return smsMessage.getSubId();
            }
        } catch (Exception e) {
            Log.e(TAG, "Error getting SIM slot", e);
        }
        return 0; // Default to SIM 1
    }

    private void sendEvent(String eventName, WritableMap params) {
        if (reactContext != null) {
            reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
        }
    }
} 

File: android/app/src/main/java/com/tesla254/Textly/SmsReceiver.kt
package com.tesla254.Textly

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Telephony
import android.telephony.SmsMessage
import com.facebook.react.ReactApplication
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule

class SmsReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (Telephony.Sms.Intents.SMS_RECEIVED_ACTION == intent.action) {
            val bundle = intent.extras
            val format = bundle?.getString("format")
            val pdus = bundle?.get("pdus") as? Array<*>

            pdus?.forEach { pdu ->
                val smsMessage = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    SmsMessage.createFromPdu(pdu as ByteArray, format)
                } else {
                    SmsMessage.createFromPdu(pdu as ByteArray)
                }

                val sender = smsMessage.displayOriginatingAddress ?: "Unknown"
                val body = smsMessage.messageBody ?: ""
                val timestamp = smsMessage.timestampMillis

                // Send event to JS
                sendEventToJS(context, sender, body, timestamp)
            }
        }
    }

    private fun sendEventToJS(context: Context, sender: String, body: String, timestamp: Long) {
        val reactApp = context.applicationContext as? ReactApplication
        val reactContext = reactApp?.reactNativeHost?.reactInstanceManager?.currentReactContext

        reactContext?.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            ?.emit("onSmsReceived", createSmsMap(sender, body, timestamp))
    }

    private fun createSmsMap(sender: String, body: String, timestamp: Long): WritableMap {
        val map = Arguments.createMap()
        map.putString("sender", sender)
        map.putString("body", body)
        map.putDouble("timestamp", timestamp.toDouble())
        return map
    }
}
